### ISSUE TITLE
JSON transcoding has backwards incompatible change in 1.27.0, due to json_name field always populated by protoc

### ISSUE THREAD

**Issue by C0mbatwombat** [line/armeria#5890]
Hi,

In #5193 functionality was added so that if a user sets the `json_name` field in the .proto file explicitly, it is used to resolve the path & query params when doing transcoding. This makes perfect sense, if it weren't for a bug in protoc: the proto compiler [**always** populates the json_name field](https://groups.google.com/g/protobuf/c/VFHoXlDbs9Y) since 3.1.0. 

This means that we have to add the actual `json_name` fields to all our proto definitions to be able to retain our existing functionality: namely to use the original field that includes underscores.

The Protobuf [code base](https://github.com/protocolbuffers/protobuf/blob/3e1967e10be786062ccd026275866c3aef487eba/src/google/protobuf/descriptor.cc#L6720) also acknowledges it as a bug and has a workaround to deal with it.

I suppose Armeria needs to use the same workaround as the protobuf codebase: if the `json_name` that is set is the same as the default calculated one, assume it was not set explicitly by the user.

Thanks!

**Comment by ikhoon** [comment]
> I suppose Armeria needs to use the same workaround as the protobuf codebase: if the json_name that is set is the same as the default calculated one, assume it was not set explicitly by the user.

It makes sense. We may provide a new option to ignore `json_name`. Let us handle it in the next version.

**Comment by ikhoon** [comment]
I've compiled a proto file with protoc 3.25.1 and `json_name` was empty if it is not set.
https://github.com/line/armeria/blob/683bd74b91eec18a1a5007c4ac1013327e7b5a62/dependencies.toml#L119
<img width="1013" alt="image" src="https://github.com/user-attachments/assets/6b0c8a56-be79-45ce-940d-4920177f8093">
Was that bug fixed in recent versions or am I missing something to reproduce the problem?

**Comment by mbgit2** [comment]
For us it doesn't seem fixed. We have a custom interceptor, but this now gets a complete empty <ReqT message> argument when using transcoding.

**Comment by berendjan** [comment]
Hi @ikhoon, 
we're also encountering the problem but only when using plugins. As noted in the [code base](https://github.com/protocolbuffers/protobuf/blob/3e1967e10be786062ccd026275866c3aef487eba/src/google/protobuf/descriptor.cc#L6720): 
_"json_name option is not allowed on extension fields. Note that the json_name field in FieldDescriptorProto is always populated by protoc when it sends descriptor data to plugins"_

If you add a plugin you will see the filled in `json_name`. 
In our setup we use bazel to compile a src jar which uses protoc plugins leading to the same error when updating to 1.27.0.

Thanks!

**Comment by ikhoon** [comment]
I failed to reproduce the problem by configuring https://github.com/bufbuild/protoc-gen-validate. Would you provide a minimal reproducible example? The example should be useful for preventing regression and ensuring compatibility.

**Comment by mbgit2** [comment]
Issue: explicitly need to set json_name in proto file to enable trancoding.

I've made a minimal [repo](https://github.com/mbgit2/armeria-example) that reproduces the issue. Note it is build using bazel, but with all previously specified versions as per this issue.

```
syntax = "proto3";

package hello.service.v1;

import "hello/v1/hello.proto";
import "google/api/annotations.proto";

option java_package = "com.example.myproject.service.hello.v1";
option java_multiple_files = true;


service HelloService {
  rpc Hello (hello.v1.HelloRequest) returns (hello.v1.HelloReply){
    option (google.api.http) = {
      get: "/v1/hello/{name_input}"
    };
  }
}

message HelloRequest {
  string name_input = 1;// [json_name="name_input"];
}

message HelloReply {
  string message = 1;
}
```

This proto compiled to java and then used by armeria:1.26 with transcoding results in  the following service printing "Hello, mbgit2!". However with all the same in armeria:1.30 it prints "Hello, !". The incoming HelloRequest is empty, unless we specify the json_name, see the commented line in the proto file. 

```
public class HelloService extends HelloServiceGrpc.HelloServiceImplBase {

    @Override
    public void hello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
        HelloReply reply = HelloReply.newBuilder()
                .setMessage("Hello, " + req.getNameInput() + '!')
                .build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
    }
}
```

```
    ServerBuilder sb = Server.builder();
    GrpcServiceBuilder grpcBuilder = GrpcService.builder();
    grpcBuilder.enableHttpJsonTranscoding(true);
    sb.service(grpcBuilder.addService(new HelloService()).build());
```

**Comment by mbgit2** [comment]
Any update would be much appreciated, would this be taken into the next version?

**Comment by ikhoon** [comment]
I managed to reproduce your repo that uses Bazel. In fact, I didn't catch the exact difference between Gradle and Bazel. Anyway, I will send a patch that will be included in the next minor version.

**Comment by C0mbatwombat** [comment]
That's good news, thanks! I understand this is a very edge-case situation, but it would help us stay updated :-)

**Comment by ikhoon** [comment]
I think this issue will be fixed by #6082. Would you like to review if the workaround resolves your problems?

**Comment by C0mbatwombat** [comment]
> > I suppose Armeria needs to use the same workaround as the protobuf codebase: if the json_name that is set is the same as the default calculated one, assume it was not set explicitly by the user.
> 
> It makes sense. We may provide a new option to ignore `json_name`. Let us handle it in the next version.

I was wondering, with your solution it is not possible to use the `json_name` field anymore at all, whereas the workaround used by protobuf does allow it to be set. Was there a reason not to go for that approach? Don't get me wrong, I'm happy that you looked at this annoying edge-case, and I don't use `json_name` anywhere (yet), so it's currently fine for me.

**Comment by ikhoon** [comment]
> I was wondering, with your solution it is not possible to use the json_name field anymore at al

Good point. I updated the PR to allow the json name field as a lower priority.
I believe that your code base will work without any modification. It will look up the json name fields first and then check the original fields. The first search does not matter since most users may have an empty or a small number of json names. 
https://github.com/line/armeria/blob/3acdb3bc14ec9c1a4f35890ce34e868a172db2c5/grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingOptionsBuilder.java#L44-L48


To get an optimized code path for your case to look up the original fields first, you can change the order of `queryParamMatchRules`. 

```java
 HttpJsonTranscodingOptions
  .builder()
  // Use only original fields
  .queryParamMatchRules(HttpJsonTranscodingQueryParamMatchRule.ORIGINAL_FIELD)
  .build()

 HttpJsonTranscodingOptions
  .builder()
  // Check the original fields first and then use json names 
  .queryParamMatchRules(HttpJsonTranscodingQueryParamMatchRule.ORIGINAL_FIELD, JSON_NAME)
  .build()
```

**Comment by C0mbatwombat** [comment]
Perfect, thank you!
